src/Microsoft.ML.Transforms/Text/NgramTransform.cs

```diff
@@ -84,64 +74,140 @@ public bool TryUnparse(StringBuilder sb)
 
             [Argument(ArgumentType.AtMostOnce,
                 HelpText = "Maximum number of tokens to skip when constructing an ngram",
                 ShortName = "skips")]
-            public int SkipLength = 0;
+            public int SkipLength = NgramEstimator.Defaults.SkipLength;
 
             [Argument(ArgumentType.Multiple, HelpText = "Maximum number of ngrams to store in the dictionary", ShortName = "max")]
-            public int[] MaxNumTerms = new int[] { DefaultMaxTerms };
+            public int[] MaxNumTerms = new int[] { NgramEstimator.Defaults.MaxNumTerms };
 
             [Argument(ArgumentType.AtMostOnce, HelpText = "The weighting criteria")]
-            public WeightingCriteria Weighting = WeightingCriteria.Tf;
+            public NgramEstimator.WeightingCriteria Weighting = NgramEstimator.Defaults.Weighting;
         }
 
-        private sealed class ColInfoEx
+        private const uint VerTfIdfSupported = 0x00010002;
+
+        public const string LoaderSignature = "NgramTransform";
+        internal const string Summary = "Produces a bag of counts of ngrams (sequences of consecutive values of length 1-n) in a given vector of keys. "
+            + "It does so by building a dictionary of ngrams and using the id in the dictionary as the index in the bag.";
+
+        internal const string UserName = "NGram Transform";
+
+        private static VersionInfo GetVersionInfo()
         {
-            // Position i, indicates whether the pool contains any (i+1)-grams
-            public readonly bool[] NonEmptyLevels;
+            return new VersionInfo(
+                modelSignature: "NGRAMTRN",
+                // verWrittenCur: 0x00010001, // Initial
+                // verWrittenCur: 0x00010002, // Add support for TF-IDF
+                verWrittenCur: 0x00010003, // Get rid of writing float size in model context
+                verReadableCur: 0x00010003,
+                verWeCanReadBack: 0x00010001,
+                loaderSignature: LoaderSignature,
+                loaderAssemblyName: typeof(NgramTransformer).Assembly.FullName);
+        }
 
+        /// <summary>
+        /// Describes how the transformer handles one column pair.
+        /// </summary>
+        public sealed class ColumnInfo
+        {
+            public readonly string Input;
+            public readonly string Output;
             public readonly int NgramLength;
             public readonly int SkipLength;
-
-            public readonly WeightingCriteria Weighting;
-
-            public bool RequireIdf()
+            public readonly bool AllLengths;
+            public readonly NgramEstimator.WeightingCriteria Weighting;
+            /// <summary>
+            /// Contains the maximum number of grams to store in the dictionary, for each level of ngrams,
+            /// from 1 (in position 0) up to ngramLength (in position ngramLength-1)
+            /// </summary>
+            public readonly int[] Limits;
```

Comment:
>int [](start = 28, length = 3)

I would make all of them readonly propertues. This particular one I would turn into `ImmutableArray<int>`